clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -855,7 +860,7 @@ private void sendProduceRequest(long now, int destination, short acks, int timeo
         if (batches.isEmpty())
             return;
 
-        final Map<TopicPartition, ProducerBatch> recordsByPartition = new HashMap<>(batches.size());
+        final Map<TopicIdPartition, ProducerBatch> recordsByPartition = new HashMap<>(batches.size());
         Map<String, Uuid> topicIds = topicIdsForBatches(batches);
```

Comment:
one question: what happens if the topic id is unknown (`Uuid.ZERO_UUID`)?

Reply:
We always fall back on topic name if topic id not passed down in the produce request and sort it out down the line when the broker recieve the produce request. It should be the same as if the client not topic id aware.

Reply:
> We always fall back on topic name if topic id not passed down

IIRC, the topic name is ignored in RPC when the version is larger than 13. So, is it possible that the broker receives the produce request without both the topic name and id?

Reply:
It shouldn't happened but if it did the batch would fail with "unkonw topic"

Reply:
The sender on client side always fetch metadata before send so we should get some sort of metadata to populate topic id

Reply:
The code has been updated to handle this potentaial of topic id zero and ensure consistent metadata between produce request and response. 🤞🏼 this would address any rare cases we might hit